### DISCUSSION TITLE
Pivots

### DISCUSSION THREAD

**Discussion by DaveSkender** [DaveSkender/Stock.Indicators#436] [Indicators]
Community discussion about Pivots.

Note, this is not the traditional [Pivot Points](https://dotnet.stockindicators.dev/indicators/PivotPoints/#content).

**Comment by DaveSkender** [comment]
The concept of pivots started as a discussion on how to visualize [divergence](https://www.investopedia.com/terms/d/divergence.asp), see #434 for the origination story of this indicator.  It is meant to be a base to enable comparing point trends any two indicators or close price and an indicator and is a more customizable variation of [Williams Fractal](https://dotnet.stockindicators.dev/indicators/Fractal/#content).

**Comment by kwcoffee1** [comment]
Hi Dave,

Can you recommend some practical values to test the Pivots indicator?
I tried: LeftSpan = 7, RightSpan = 3, MaxTotal = 30
Can you command on the purpose and the assigned values of these parameters?
Also the description of the 3 returned value, like HighLine?

Thanks

**Reply by DaveSkender** [reply]
![image](https://github.com/DaveSkender/Stock.Indicators/assets/8432125/ee8336ac-f561-474f-be9d-0738d5c755d4)

Pivots are simply a more customizable version of a [Williams Fractal](https://dotnet.stockindicators.dev/indicators/Fractal/#content).  Based on his own analysis, Williams preferred these settings:

`leftSpan = 2`
`rightSpan = 2`
`endType = high/low`

To elaborate on the [description in our documentation site](https://dotnet.stockindicators.dev/indicators/Pivots/#content), here's a bit more of an explanation and meaning.

### the pivots

The pivots need `leftSpan`, `rightSpan`, and `endType` for calculation and represent a change in direction of price.

![image](https://github.com/DaveSkender/Stock.Indicators/assets/8432125/1b060325-9ce7-48c0-8fad-f6cc5a1ff402)

When a point pivot is detected, there will be a `HighPoint` or `LowPoint` value depicting the actual corresponding high or low point price.

Use larger span values to depict larger trends, but keep in mind that this is a retrospective analysis that will never identify a point until `rightSpan` periods have occurred, for confirmation.

### the trend markers

To determine the trend markers (HH, LH, LL, HL) for an identified point, we look backwards `maxTrendPeriods` number of periods (bars) to find the last point.  If the current point is a high point and is higher than the last high point, it's marked as a higher high (HH) in the `HighTrend` field.  We do the same equivalent markings for lower high (LH) and for low point trends in the `LowTrend` field.  We separate these low and high fields because a) there could be both a high/low point in the same bar in rare circumstances and 2) it's easier to consume for charting use cases.

### the trend lines

`HighLine` and `LowLine` are simple values representing the line value between corresponding trend markers.  Absence of a trendline value means the two adjacent trend markers were more than `maxTrendPeriods` away from each other and were not attributable to each other.  This is also why we can't identify the marker of the first point in a contiguous trendline.  Setting a very high `maxTrendPeriods` will produce more unbroken trendlines, but it comes at a cost of an exponentially slower computed analysis.

**Comment by kwcoffee1** [comment]
Thank you very much for the quick response.

I notice the default uses a maxTrendPeriods of 20 "periods". Is this the number of the 5-bar pivot evaluation periods?
That is what does maxTrendPeriods denotes?

After investigating the data furthermore, I believe maxTrendPeriods is the actual bar count and we should filter on the chances of the Trend and not on the individual bars as indicate return — your explanation sounds right.

**Reply by DaveSkender** [reply]
I used the term "periods" to mean the same thing as bars or candles.  For example, if you're using daily quotes, 5 periods = 5 days.  `maxTrendPeriods` is only used to limit the lookback period when identifying trends.

**Comment by Paroca72** [comment]
Hello,

I'm using 26 as period but the result are more that 1400 items.
I expected something like less 26 items as result.

<img width="1243" height="355" alt="image" src="https://github.com/user-attachments/assets/3e1b0796-0304-4f54-b15d-80d05ad869a5" />

What I'm missing?

Thanks for your support
Sam

**Reply by DaveSkender** [reply]
Are there actually 1400 items in `candles`?

**Reply by DaveSkender** [reply]
> Yes. the other indicators work perfect with the same source of candles.

upload your raw `candles` and `Period` values to a [new issue](https://github.com/DaveSkender/Stock.Indicators/issues/new) and I’ll try to repro.

**Reply by DaveSkender** [reply]
@Paroca72 I've added some suggestions for your code in this issue, and also updated [docs](https://dotnet.stockindicators.dev/indicators/Pivots/#parameters) to clarify what `maxTrendPeriods` means -- it's certainly confusing.

- #1950

**Comment by Paroca72** [comment]
Yes.. 
the other indicators work perfect with the same source of candles.
